File: maple-engine-plugins/maple-spark/apps/maple-spark-data-calc/src/main/scala/org/xi/maple/datacalc/spark/api/MapleSink.scala
package org.xi.maple.datacalc.spark.api

import org.apache.commons.lang3.StringUtils
import org.apache.spark.sql.{Dataset, Row}
import org.xi.maple.common.util.VariableUtils
import org.xi.maple.datacalc.spark.model.SinkConfig

import java.util

trait MapleSink[T <: SinkConfig] extends MaplePlugin[T] with Logging {

  override protected def getData(variables: util.Map[String, String]): Dataset[Row] = {
    val ds: Dataset[Row] = if (StringUtils.isBlank(config.getSourceQuery)) {
      spark.read.table(config.getSourceTable)
    } else {
      val sourceQuery = VariableUtils.replaceVariables(config.getSourceQuery, variables)
      spark.sql(sourceQuery)
    }
    val partitions = config.getNumPartitions
    if (partitions != null && partitions > 0) {
      ds.repartition(partitions)
    } else {
      ds
    }
  }
}

File: maple-engine-plugins/maple-flink/apps/maple-flink-data-calc/src/main/java/org/xi/maple/datacalc/flink/api/MapleSink.java
package org.xi.maple.datacalc.flink.api;

import org.apache.commons.lang3.StringUtils;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.TableDescriptor;
import org.apache.flink.table.api.TableEnvironment;
import org.apache.flink.table.api.TablePipeline;
import org.apache.flink.table.catalog.Column;
import org.xi.maple.datacalc.flink.exception.ConfigRuntimeException;
import org.xi.maple.datacalc.flink.model.SinkConfig;
import org.xi.maple.datacalc.flink.util.TableUtils;

import java.util.List;

public abstract class MapleSink<T extends SinkConfig> extends MaplePlugin<T> implements TableDefine {

    final Table sourceTable;

    public MapleSink(TableEnvironment tableEnv) {
        super(tableEnv);
        sourceTable = StringUtils.isNotBlank(config.getSourceQuery())
                ? tableEnv.sqlQuery(config.getSourceQuery())
                : tableEnv.from(config.getSourceTable());
    }

    @Override
    public void define() {
        TableDescriptor tableDescriptor = TableUtils.getTableDescriptor(config);
        tableEnv.createTemporaryTable(config.getResultTable(), tableDescriptor);
    }

    public TablePipeline getTablePipeline() {
        return sourceTable.insertInto(config.getResultTable());
    }

    @Override
    public void prepare() {
        List<Column> columns = sourceTable.getResolvedSchema().getColumns();
        int columnSize = config.getPhysicalColumns().size();
        if (!config.getMetadataColumns().isEmpty()) {
            columnSize += (int) config.getMetadataColumns().stream().filter(item -> !item.isVirtual()).count();
        }
        if (columns.size() != columnSize) {
            throw new ConfigRuntimeException("The number of columns in the source table and the target table is inconsistent");
        }
    }
}
